File: week4/2_hash_chains/hash_chains.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

struct Query {
    string type, s;
    size_t ind;
};

class QueryProcessor {
    int bucket_count;
    // store all strings in one vector of unordered sets. Won't maintain insertion order
    vector<unordered_set<string>> sets;

    size_t hash_func(const string& s) const {
        static const size_t multiplier = 263;
        static const size_t prime = 1000000007;
        unsigned long long hash = 0;
        for (int i = static_cast<int> (s.size()) - 1; i >= 0; --i)
            hash = (hash * multiplier + s[i]) % prime;
        return (hash % bucket_count + bucket_count) % bucket_count;
    }

public:
    explicit QueryProcessor(int bucket_count): bucket_count(bucket_count) {
        sets.resize(bucket_count);
    }

    static Query readQuery() {
        Query query;
        cin >> query.type;
        if (query.type != "check")
            cin >> query.s;
        else
            cin >> query.ind;
        return query;
    }

    static void writeSearchResult(bool was_found) {
        std::cout << (was_found ? "yes\n" : "no\n");
    }

    void processQuery(const Query& query) {
        if (query.type == "check") {
            unordered_set<string> & theSet = sets[query.ind];
            for (const string& s : theSet) {
                std::cout << s << " ";
            }
            std::cout << "\n";

        } else {
            unsigned long long hashVal = hash_func(query.s);
            unordered_set<string> & theSet = sets[hashVal];
            auto iter = theSet.find(query.s);
            if (query.type == "find")
                writeSearchResult(iter != theSet.end());
            else if (query.type == "add") {
                if (iter == theSet.end()) {
                    auto iterBegin = theSet.begin();
                    theSet.emplace_hint(iterBegin, query.s);
                }
            } else if (query.type == "del") {
                if (iter != theSet.end())
                    theSet.erase(iter);
            }
        }
    }

    void processQueries() {
        int n;
        cin >> n;
        for (int i = 0; i < n; ++i)
            processQuery(readQuery());
    }
};

int main() {
// This won't pass the assignment. Unlike Python, C++ maps or sets don't keep insertion order
    std::ios_base::sync_with_stdio(false);
    int bucket_count;
    cin >> bucket_count;
    QueryProcessor proc(bucket_count);
    proc.processQueries();
    return 0;
}
/*
 5
 12
 add world
 add HellO
 check 4
 find World
 find world
 del world
 check 4
 del HellO
 add luck
 add GooD
 check 2
 del good
 */

File: week4/2_hash_chains/hash_chains.py
# python3


class Query:
    def __init__(self, query):
        self.type = query[0]
        if self.type == 'check':
            self.ind = int(query[1])
        else:
            self.s = query[1]


class QueryProcessor:
    _multiplier = 263
    _prime = 1000000007

    def __init__(self, bucket_count):
        self.bucket_count = bucket_count
        # store strings in dictionaries inside list based on their hash-key
        self.bucket = [{} for _ in range(self.bucket_count)]

    def _hash_func(self, s):
        ans = 0
        for c in reversed(s):
            ans = (ans * self._multiplier + ord(c)) % self._prime
        return ans % self.bucket_count

    def write_search_result(self, was_found):
        print('yes' if was_found else 'no')

    def write_chain(self, chain):
        print(' '.join(chain))

    def read_query(self):
        return Query(input().split())

    def process_query(self, query):
        if query.type == "check":
            # use reverse order, because we append strings to the end of dict
            self.write_chain(reversed(list(cur for cur in (self.bucket[query.ind]))))
        else:
            key = self._hash_func(query.s)
            if query.type == 'find':
                if query.s in self.bucket[key].keys():
                    self.write_search_result(True)
                else:
                    self.write_search_result(False)
            elif query.type == 'add':
                if query.s not in self.bucket[key].keys():
                    self.bucket[key].update({query.s: None})
            else:  # if delete
                if query.s in self.bucket[key].keys():
                    self.bucket[key].pop(query.s)

    def process_queries(self):
        n = int(input())
        for i in range(n):
            self.process_query(self.read_query())


if __name__ == '__main__':
    buckets = int(input())
    proc = QueryProcessor(buckets)
    proc.process_queries()

'''
 5
 12
 add world
 add HellO
 check 4
 find World
 find world
 del world
 check 4
 del HellO
 add luck
 add GooD
 check 2
 del good
 
3
12
check 0
find help
add help
add del
add add
find add
find del
del del
find del
check 0
check 1
check 2
'''
